include/cliffordsynthesis/CliffordSynthesizer.hpp

```diff
@@ -0,0 +1,289 @@
+#include <utility>
+#include <vector>
+
+class CliffordOptimizer {
+public:
+    CliffordOptimizer() = default;
+    void init(bool pchoose_best = false, bool puse_embedding = false, unsigned char pnqubits = 0, unsigned short pinitial_timesteps = 0, SynthesisStrategy strategy = SynthesisStrategy::UseMinimizer, SynthesisTarget ptarget = SynthesisTarget::GATES);
+    void setCircuit(const qc::QuantumComputation& qc);
+    void setTableau(Tableau& targetTabl);
+    void optimize();
+
+    void setArchitecture(const Architecture& arch) {
+        architecture = arch;
+        if (nqubits == 0) {
+            nqubits = architecture.getNqubits();
+        }
+        auto map = highestFidelityMap.emplace_back();
+        architecture.getHighestFidelityCouplingMap(nqubits, map);
+    }
+    bool                   choose_best   = false;
+    bool                   use_embedding = false;
+    unsigned char          nqubits       = 0U;
+    std::set<signed char>  used_qubits{};
+    unsigned short         initial_timesteps = 0U;
+    int                    verbose           = 0;
+    int                    nthreads          = 1;
+    SynthesisStrategy      strategy          = SynthesisStrategy::UseMinimizer;
+    SynthesisTarget        target            = SynthesisTarget::GATES;
+    SynthesisMethod        method            = SynthesisMethod::Z3;
+    qc::QuantumComputation circuit;
+
+    std::vector<CouplingMap> highestFidelityMap;
+
+    Tableau initialTableau{};
+    Tableau targetTableau{};
+    Tableau modelTableau{};
+
+    void dumpResult(const std::string& outputFilename) {
+        if (optimal_results.resultCircuit.empty()) {
+            std::cerr << "Circuit is empty." << std::endl;
+            return;
+        }
+
+        size_t      dot       = outputFilename.find_last_of('.');
+        std::string extension = outputFilename.substr(dot + 1U);
+        std::transform(extension.begin(), extension.end(), extension.begin(), [](unsigned char c) { return ::tolower(c); });
+        if (extension == "real") {
+            dumpResult(outputFilename, qc::Real);
+        } else if (extension == "qasm") {
+            dumpResult(outputFilename, qc::OpenQASM);
+        }
+    }
+
+    void dumpResult(const std::string& outputFilename, qc::Format format) {
+        optimal_results.resultCircuit.dump(outputFilename, format);
+    }
+
+    void dumpResult(std::ostream& os, qc::Format format) {
+        optimal_results.resultCircuit.dump(os, format);
+    }
+
+    CliffordOptimizationResults optimal_results{};
+
+protected:
+    Architecture                architecture{};
+    CliffordOptimizationResults main_optimization(
+            int                                                        timesteps,
+            const std::set<std::pair<unsigned short, unsigned short>>& reducedCM,
+            const std::vector<unsigned short>& qubitChoice, Tableau& initialTableau,
+            Tableau& targetTableau);
+
+    void make_depth_optimizer(
```

Comment:
Seeing these three functions and the whole code in this file I am wondering. Would it actually make sense to create a base class for the optimizer and then provide three derived classes (in separate files) for the different objective functions. It would certainly clean up the library here and would act as a separation of concerns.

Reply:
This is something thats in my todos, but something thats a little bit less priority than getting the PR merge ready at the moment.

Reply:
To be honest, I believe that change is part of the PR being merge-ready. At the moment, the structure in this file is a bit of a mess that's particularly hard to review/grasp.
I guess it's best to resolve the other remarks first, but I really see this as an important part to make all of this seem a little less "hacky".

Reply:
I've tried putting the different Targets (Gates, Depth, Fidelity) in different files, but found that it does not improve readability of the CliffordSynthesizer class that much, as it only removes three methods effectively. 
the main complexity still remains (with single and multi gate constraints and the assertion of the tableau)